### ISSUE TITLE
Display Images using the Kitty graphics protocol

### ISSUE THREAD

**Issue by makzsslm** [rigormorrtiss/discordo#476]
Are there any plans to implement the [Kitty graphics protocol](https://sw.kovidgoyal.net/kitty/graphics-protocol/) to display gifs, images, et al. ?

**Comment by youmukonpaku1337** [comment]
what about img2sixel so that there is more compatibility? one of the main usecases i see this for is running on weaker/embedded devices

**Comment by makzsslm** [comment]
[Looks interesting](https://github.com/saitoha/libsixel), but the last commit was 6 years ago !

**Comment by youmukonpaku1337** [comment]
> [Looks interesting](https://github.com/saitoha/libsixel), but the last commit was 6 years ago !

if it aint broke don't fix it 😎

**Comment by plasmoduck** [comment]
Yes I would love to see inline images. Instead of showing the image link, have discordo display the image using the kitty icat kitten - https://sw.kovidgoyal.net/kitty/kittens/icat/

**Comment by linuxmobile** [comment]
Any way to work on this?

**Comment by radu-corbu** [comment]
Just wanted to chime in and say this would be a really neat feature to have. Just came across Discordo and I would very much like to use it as my main Discord client. However, the lack of in-client images is a bit of a turn-off. 

I understand  there are limitations to what you can do inside of a terminal window, but image rendering seems more and more doable nowadays. For example, Yazi can render [image previews](https://yazi-rs.github.io/docs/image-preview/) just fine, without much setup on my part. 

Looking forward to seeing this feature added to Discordo!

**Comment by couleurm** [comment]
what about first trying with kitty images, with a fallback to sixel/low resolution ascii?

**Comment by Amr-Shams** [comment]
i have tried with kitty it works super fine, soon i will create a pull request, (if the client/terminal doesn't support the format) the image will be a blank (also will not work on wsl easily)

**Comment by ossenthusiast** [comment]
> i have tried with kitty it works super fine, soon i will create a pull request, (if the client/terminal doesn't support the format) the image will be a blank (also will not work on wsl easily)

🥳 Sounds good, thank you very much!

**Comment by linuxmobile** [comment]
I've started working on the PR to include image previews, but I am finding it quite difficult to add support for images in `tview`. However, for now I have temporary support for ASCII. I will continue trying to implement a high-quality solution with sixel, kitty-image-protocol, and iterm2.

<img width="947" height="1064" alt="Image" src="https://github.com/user-attachments/assets/52ce9335-fae0-45d0-bf55-b08c6e7d9682" />

EDIT: ascii work fine in tmux or zellij.

**Comment by Amr-Shams** [comment]
This is much better progress than mine tbh

**Comment by linuxmobile** [comment]
> This is much better progress than mine tbh

I'm dealing with the implementation of image preview in tview, but it's working:

<img width="638" height="565" alt="Image" src="https://github.com/user-attachments/assets/285b24aa-c5f3-478d-9a2b-bc23f6394b4c" />

**Comment by ossenthusiast** [comment]
Thanks for the update! Nice work getting the ASCII preview in. Image support in tview is definitely non-trivial, so that makes sense. Looking forward to seeing what you come up with with sixel/kitty/iTerm2. Also, feel free to join the Discord if you’d like to discuss ideas or progress there: https://discord.com/invite/VzF9UFn2aB

**Comment by ghost** [comment]
A while ago I added kitty support (it can do sixels as well) to mpv because I’m using it in the moderation TUI for my temp file hoster project. It can be used via libmpv or mpv’s JSON IPC, and supports video playback.

To do this, you set at least the following options (do read the manual!):

```
profile=sw-fast
vo=kitty
vo-kitty-config-clear=no
vo-kitty-alt-screen=no
```

Unfortunately, it does not support more than one image or video at a time because in mpv, the video output driver cannot read back replies (like ID assignments) from the terminal.

You could work around this by buffering mpv’s output, and, whenever you encounter the kitty end sequence `\033\\`, positioning the cursor at the desired position and then writing the buffer to the terminal. You will also need to filter out the clear-all-images sequence `\033_Ga=d;`.

**Comment by TechyTrickster** [comment]
I would also like to see image support in here.  My understanding (I could be quite wrong) is that sixel support would be the minimum viable product, as its more widely supported.  https://www.arewesixelyet.com vs https://sw.kovidgoyal.net/kitty/graphics-protocol/

**Comment by whazlewo** [comment]
I got a POC working using Claude Code a while back just out of curiosity. Kitty seemed to work fine over ssh using ghostty, but then I just tossed it because I didn't have time to go review precisely how it had been implemented.  I hope someone has time to do this correctly. It's the only way I currently have to access Discord at all.

**Comment by Arzumify** [comment]
I think the best approach to this would be to fork https://github.com/BourgeoisBear/rasterm/ and add proper sixel support (probably via https://github.com/mattn/go-sixel/blob/master/sixel.go). This + a convenience interface that abstracts away the actual image would probably make it pretty trivial.

I'll try an implementation.

**Comment by Arzumify** [comment]
> I think the best approach to this would be to fork https://github.com/BourgeoisBear/rasterm/ and add proper sixel support (probably via https://github.com/mattn/go-sixel/blob/master/sixel.go). This + a convenience interface that abstracts away the actual image would probably make it pretty trivial.
> 
> I'll try an implementation.

package is under development: https://github.com/Arzumify/pritty
they are all working, i want to rewrite the API for the wezterm backend to be a bit neater, as well as add some helper functions that can automatically decide which to use and abstract over all 3 backends such that the caller doesn't have to worry
